FEXCore: Reclaimable thread pool allocator

Sonicadvance1 opened the pull request:
Creates a pool allocator for OpcodeDispatcher and IRCompaction that
shares memory allocations between threads in a pool and supports
reclaiming stale allocations from participating threads.

A thread will use a heuristic to keep its claimed memory allocation
around if it is allocating a lot of code. If it slows down then it will
start putting the memory allocation back in to the thread pool.

Additionally if the allocation has been "disowned" and gone to sleep
while still retaining the allocation, then another thread can inspect
 these stale allocations and reclaim it from the idling thread. Saving
further memory.

This needs some more work and cleanup but this is an interesting concept
that saves a decent amount of memory even in a basic test.

Causes teeworlds' title screen to go from 754MB to 599MB in my simple
test. 79.4% the memory usage is a good start.

First step for #1625

neobrain commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,160 @@
+        std::atomic<bool> *CurrentClientOwnedFlag{};
+      };
+
+      using ContainerType = std::list<MemoryBuffer*>;
```
Would `std::vector` work here? Looks like we're mostly just iterating over these (where `vector` provides better cache locality) and insertion/erasure is exceptional.

Sonicadvance1 replied:
I very explicitly chose list here so iterators aren't invalidated on insert/delete.
Could add a comment stating why I chose list.

Sonicadvance1 replied:
Added a comment that I'm requiring iterators to stay valid while manipulating the container. Resolved.

neobrain commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,160 @@
+
+      ~ThreadPoolAllocator() {
+        // Cleanup any buffers that are remaining
+        for (auto it = Buffers.begin(); it != Buffers.end(); ++it) {
```
range-based for loops can be used here and in a couple of other places.

Sonicadvance1 replied:
Range based used for the ones that can. The other ones still need the iterators to be passed around.

neobrain commented on External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:
```diff
@@ -20,19 +21,12 @@ namespace FEXCore::IR {
  * Can potentially support reallocation if we are smart and make sure to invalidate anything holding a true pointer
  */
-class DualIntrusiveAllocator final {
+class DualIntrusiveAllocator {
```
Should this have a virtual destructor? If not, we should use non-public inheritance instead so people can't accidentally bind child objects to `DualIntrusiveAllocator*` pointers.

Sonicadvance1 replied:
Yes. Resolved.

neobrain replied:
What happened to this? Looks like the destructor is gone again now.

neobrain commented on External/FEXCore/include/FEXCore/IR/IREmitter.h:
```diff
@@ -19,11 +19,37 @@ friend class FEXCore::IR::Pass;
+      DualListData.ClaimBuffer();
+    }
+    void UnclaimBuffer() {
+      auto Now = std::chrono::high_resolution_clock::now();
```
Would a `steady_clock` be sufficient here as well? A high resolution clock doesn't appear necessary given that the reclamation interval is 5 seconds, and `steady_clock` (being a monotonic clock) will have more predictable behavior on system clock changes. (You wouldn't want reclamation to stop working for an hour because the system entered DST!)

Sonicadvance1 replied:
You right. Resolved.

neobrain commented on External/FEXCore/include/FEXCore/IR/IntrusiveIRList.h:
```diff
@@ -20,19 +21,12 @@ namespace FEXCore::IR {
-class DualIntrusiveAllocator final {
+class DualIntrusiveAllocator {
   public:
     DualIntrusiveAllocator() = delete;
```
Aren't these two constructors implicitly deleted due to providing the `size_t` constructor?

Sonicadvance1 replied:
You right. Resolved.

Sonicadvance1 commented:
Fixed the remaining bugs in this and added a bunch of comments about how to use the API.
Now the most major contributor to memory consumption is the register allocator. Specifically the graph nodes since they aren't ever freed.

Once this is merged those can be looked at to fix its memory consumption, since it isn't quite as straightforward as just moving the allocator over to using these.

Sonicadvance1 commented:
![Image_2022-04-24_11-30-29](https://user-images.githubusercontent.com/1018829/164991172-4d1e347c-dcf4-467c-862c-41cb6f82012e.png)
Heaptrack shows a significant change but it isn't this significant because some allocations have switched from malloc to mmap. 1.6G down to 229.3MB within FEXLoader.
I don't think there is a tool that can see resident memory usage reduction other than what tools like htop give, sadly.

skmp commented:
I'll get on reviewing this first thing tomorrow, sorry for the delay

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+    public:
+      ThreadPoolAllocatorMalloc() = default;
+
+      virtual ~ThreadPoolAllocatorMalloc() {
```
Since both dtors `ThreadPoolAllocatorMalloc`, `ThreadPoolAllocatorMMap` are the same, wouldn't it be nicer to have the dtor logic in `IntrusiveThreadPoolAllocator` and rely on the alloc/free overloads? Unless I'm missing something

Sonicadvance1 replied:
Can't
`Call to pure virtual member function 'Free' has undefined behavior; overrides of 'Free' in subclasses are not available in the destructor of 'IntrusivePooledAllocator'  ThreadPoolAllocator.h:323:7: note: 'Free' declared here [-Wcall-to-pure-virtual-from-ctor-dtor]`

skmp replied:
Doh, right, derived class is destructed at that point. How about moving them to a function so that we don't have two copies of the code around?

Sonicadvance1 replied:
Moved duplicated freeing logic to its own function. Resolved now.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+            if (LastUsed != BuffersEnd) {
+              Free((*LastUsed)->Ptr, (*LastUsed)->Size);
+              delete *LastUsed;
+              OwnedBuffers.erase(LastUsed);
```
`it = OwnedBuffers.erase(LastUsed);` ?

Sonicadvance1 replied:
This is no longer isn't the for loop, there is no iterator to update.

skmp replied:
Misread the indentation. So at most one buffer gets free'd for every call to `ClaimBufferImpl`. Is this to avoid stutter?

Sonicadvance1 replied:
Yes, only one unclaimed buffer gets freed and removed from the `UnclaimedBuffers` container each time a new claim comes in. This is setup this way for longer running applications having mostly idle threads and then a handful of active threads will slowly chip away at the problem.
Also if you have an intermediate mix of active and inactive, you aren't suddenly getting a spike of deleted unclaimed buffers and new threads having absolutely no buffers available.

Basically the five second buffer of only freeing one buffer at a time is a nice sponge in both directions.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+            BestFit = UnsizedFit;
+          }
+
+          {
```
```suggestion
          // Free owned buffers that have expired
          {
```

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+            ++it;
+          }
+        }
+
```
```suggestion
          // Find an owned buffer that is >= Size and Free owned buffers that have expired
```

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+        ContainerType::iterator BestFit = BuffersEnd;
+        ContainerType::iterator UnsizedFit = BuffersEnd;
+
+        auto Now = ClockType::now();
```
```suggestion
        auto Now = ClockType::now();
        // Move any expired NonOwnedBuffers to OwnedBuffers
```

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+      /**
+       * @brief List of buffers that this pool allocator itself owns
+       */
+      ContainerType OwnedBuffers;
```
How about renaming to `UnclaimedBuffers`,  and `ClaimedBuffers` so that `ClaimBuffer`/`ClaimBufferImpl` & `UnclaimBuffer`/`UnclaimBufferImpl` are more intuitive to understand?

`DisownBuffer` and `ReownOrClaimBuffer` don't modify those lists, and from my understanding they track the `claimed`/`unclaimed` statuses, not `owned`/`disowned`

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+   *    - Frees stale buffers opportunistically
+   */
+  template<typename Type, size_t PeriodMS, size_t PeriodFrequency>
+  class FixedSizeDelayedObjectReclaimer final {
```
How about we name this from `FixedSizeDelayedObjectReclaimer` to `FixedSizePooledAllocation` for clarity?

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,507 @@
+   *
+   * During buffer reclaiming is also when unclaimed buffers get freed. This means active threads are able to clean up idle thread's unused memory.
+   */
+  class IntrusiveThreadPoolAllocator {
```
How about we rename this to  `IntrusivePooledAllocator` to make it clear it doesn't allocate from a `ThreadPool`?

Sonicadvance1 replied:
Resolved.

skmp commented:
Some performance points for discussion:

Switching `IntrusiveThreadPoolAllocator::ContainerType` to a multimap structure, then ordering owned buffers by size can make finding the right buffer O(ln) from O(n), possibly a major win depending on how many buffers we have in flight.

Not sure the same idea would be applicable for unowned buffers (sorted by last used time), as that'd require a re-insertion on disown which will increase the cost there.

What is the cost of `std::chrono` and such? Do they result in syscalls? Can we guarantee somehow `rdtsc`/`cntvct_el0` for it?

Sonicadvance1 commented:
`std::chrono` doesn't result in a syscall. It'll use the cycle counter registers, potentially through a vdso. quite lightweight.

I haven't noticed any performance concerns from the current `std::list` container but I've also not found something that pedantically disowns thousands of threads worth of buffers and then spins all of them up simultaneously so they are all hunting for buffers to reown.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,502 @@
+            BestFit = UnsizedFit;
+          }
+
+          // Free owned buffers that have expired
```
```suggestion
          // Free up to one unclaimed buffer that has expired
```

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,502 @@
+          }
+        }
+
+        // Find an owned buffer that is >= Size and Free owned buffers that have expired
```
```suggestion
        // Find an unclaimed buffer that is >= Size and Free up to one unclaimed buffer that has expired
```

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,502 @@
+  /**
+   * @brief Thread pool allocator that allocates and frees objects that uses mmap
+   */
+  class ThreadPoolAllocatorMMap final : public IntrusivePooledAllocator {
```
```suggestion
  class PooledAllocatorMMap final : public IntrusivePooledAllocator {
```

or 

```suggestion
  class PooledMMapAllocator final : public IntrusivePooledAllocator {
```

Sonicadvance1 replied:
Resolved.

skmp commented on External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -0,0 +1,502 @@
+  /**
+   * @brief Thread pool allocator that allocates and frees objects using malloc
+   */
+  class ThreadPoolAllocatorMalloc final : public IntrusivePooledAllocator {
```
```suggestion
  class PolledAllocatorMalloc final : public IntrusivePooledAllocator {
```

Or

```suggestion
  class PolledMallocAllocator final : public IntrusivePooledAllocator {
```
?

skmp reviewed (approved):
Noted a few naming nitpicks, Looks good overall

skmp commented:
> `std::chrono` doesn't result in a syscall. It'll use the cycle counter registers, potentially through a vdso. quite lightweight.
> 
> I haven't noticed any performance concerns from the current `std::list` container but I've also not found something that pedantically disowns thousands of threads worth of buffers and then spins all of them up simultaneously so they are all hunting for buffers to reown.

More of a note for the future then
